### PR TITLE
[SPARK-36463][SS] Prohibit update mode in streaming aggregation with session window

### PR DESCRIPTION
**Pull request by HeartSaVioR**
### What changes were proposed in this pull request?

This PR proposes to prohibit update mode in streaming aggregation with session window.

UnsupportedOperationChecker will check and prohibit the case. As a side effect, this PR also simplifies the code as we can remove the implementation of iterator to support outputs of update mode.

This PR also cleans up test code via deduplicating.

### Why are the changes needed?

The semantic of "update" mode for session window based streaming aggregation is quite unclear.

For normal streaming aggregation, Spark will provide the outputs which can be "upsert"ed based on the grouping key. This is based on the fact grouping key won't be changed.

This doesn't hold true for session window based streaming aggregation, as session range is changing.

If end users leverage their knowledge about streaming aggregation, they will consider the key as grouping key + session (since they'll specify these things in groupBy), and it's high likely possible that existing row is not updated (overwritten) and ended up with having different rows.

If end users consider the key as grouping key, there's a small chance for end users to upsert the session correctly, though only the last updated session will be stored so it won't work with event time processing which there could be multiple active sessions.

### Does this PR introduce _any_ user-facing change?

No, as we haven't released this feature.

### How was this patch tested?

Updated tests.

### CODE REVIEW

**Comment by HeartSaVioR** [comment]
To respect the semantic of "update" properly, we need "retraction" to provide two different events, remove old session and insert updated session. This could be implemented manually with flatMapGroupsWithState, though I'm not sure this is performant enough in practice, since it requires two operations "delete" and "insert" against external storage.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingSessionWindowSuite.scala [hunk 1]
```diff
@@ -75,21 +76,7 @@ class StreamingSessionWindowSuite extends StreamTest
...
-
-    val sessionUpdates = events
-      .groupBy(session_window($"eventTime", "10 seconds") as 'session, 'sessionId)
-      .agg(count("*").as("numEvents"))
-      .selectExpr("sessionId", "CAST(session.start AS LONG)", "CAST(session.end AS LONG)",
-        "CAST(session.end AS LONG) - CAST(session.start AS LONG) AS durationMs",
-        "numEvents")
-
-    sessionUpdates.explain()
+    val sessionUpdates = sessionWindowQuery(inputData)
```
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingSessionWindowSuite.scala:79]: The most changes on this suite are about deduplication on queries. We can simply use two queries (keyed window vs global window) regardless of output mode.

**Review by viirya — COMMENTED** [review]
I think we also need to update (or mention it) in the document.

**Review by viirya — COMMENTED** [review]
Looks okay to me. I will look this again after the document is changed too.

**Review by viirya — APPROVED** [review]

**Comment by viirya** [comment]
> This doesn't hold true for session window based streaming aggregation. If you're trying to upsert the output based on the grouping key, it's high likely possible that existing row is not updated (overwritten) and ended up with having different rows.

To be clear, I looked the current approach of UPDATE mode. We considers the row is updated if there is no existing row with the state key (session key + session start time), or the stored value isn't the same as the current value that will be stored.

But it is also possible that the session window was extended backward at the session start time. So the updated rows are not actually accurate.

**Comment by HeartSaVioR** [comment]
I think the semantic is meaningful only when end users can store the output correctly. That said, we should evaluate the semantic in point of **end users' view**. They wouldn't know whether they need to see the grouping key as `grouping key` vs `grouping key + session`. `grouping key + session start` is something Spark internally uses as state key, which end users wouldn't know, so no meaning in point of end users' view.

If they leverage their knowledge about streaming aggregation, they will consider the key as `grouping key + session` (since they'll specify these things in `groupBy`) which I already demonstrated the problem.

If they consider the key as `grouping key`, there's a chance for end users to upsert the session correctly, though only the last updated session will be stored, so it won't work with event time processing.

**Comment by viirya** [comment]
Thanks for updating the description. lgtm

**Comment by HeartSaVioR** [comment]
Thanks again @viirya for the quick reviewing! Merging to master/3.2.

**Pull request closed without merging**